Backends/PDF/sheet.lisp

```diff
@@ -129,41 +116,85 @@
+(defmethod initialize-instance :after ((graft pdf-graft) &key)
+  (setf (slot-value graft 'native-transformation) nil)
+  (setf (slot-value graft 'native-region) nil))
 
 (defun graft-length (length units)
   (* length (ecase units
-              (:device       (/ 720 254))
-              (:inches       (/ 10 254))
-              (:millimeters  1)
+              (:device       1)
+              (:inches       (/ 72))
+              (:millimeters  (/ 254 720))
               (:screen-sized (/ length)))))
 
 (defmethod graft-width ((graft pdf-graft) &key (units :device))
-  (graft-length (pdf-graft-width graft) units))
+  (graft-length (bounding-rectangle-width (sheet-native-region graft)) units))
 
 (defmethod graft-height ((graft pdf-graft) &key (units :device))
-  (graft-length (pdf-graft-height graft) units))
-
-(defun make-pdf-graft ()
-  (make-instance 'pdf-graft))
+  (graft-length (bounding-rectangle-height (sheet-native-region graft)) units))
 
 (defmethod sheet-region ((sheet pdf-graft))
   (let ((units (graft-units sheet)))
     (make-rectangle* 0 0
                      (graft-width sheet :units units)
                      (graft-height sheet :units units))))
 
+
+(defmethod sheet-native-region ((sheet pdf-graft))
+  (with-slots (native-region) sheet
+    (unless native-region
+      (setf native-region
+            (paper-region (device-type (port sheet))
+                          (page-orientation (port sheet)))))
+    native-region))
+
+;; This is necessary because McCLIM didn't reset the
+;; native-transformation for basic-sheet. -- admich 2020-01-30
+(defmethod invalidate-cached-transformations ((sheet pdf-graft))
+  (with-slots (native-transformation device-transformation) sheet
+    (setf native-transformation nil
+     device-transformation nil))
+  (loop for child in (sheet-children sheet)
+     do (invalidate-cached-transformations child)))
+
+(defmethod graft-units-transformation ((graft pdf-graft))
+  (ecase (graft-units graft)
+    (:device +identity-transformation+)
+    (:inches (make-scaling-transformation* 72 72))
+    (:millimeters (make-scaling-transformation* (/ 720 254) (/ 720 254)))
+    (:screen-sized (make-scaling-transformation* (graft-width graft) (graft-height graft)))))
+
+(defmethod graft-orientation-transformation ((graft pdf-graft))
+  (ecase (graft-orientation graft)
+    (:graphics +identity-transformation+)
+    (:default (compose-transformations
+               (make-translation-transformation
+                0
+                (bounding-rectangle-height (sheet-native-region graft)))
+               (make-reflection-transformation* 0 0 1 0)))))
+
+(defmethod sheet-native-transformation ((sheet pdf-graft))
+  (with-slots (native-transformation) sheet
+    (unless native-transformation
+      (setf native-transformation
+            (compose-transformations
+             (graft-orientation-transformation sheet)
+             (graft-units-transformation sheet))))
+    native-transformation))
```

Comment:
please define graft-*-transformation functions with defun, it is a local functionality which is not exported, so adding implicit generic functions doesn't make sense now.

When we sort out how we should handle transformations (the other ticket) this protocol may be necessary and then it will land in clime or climb package, but for now there is no benefit of that.

Reply:
If you agree with my reply to the previous request I make this changing, otherwise I will work a little more on the postscript backend.

Reply:
ditto, please make it defun and duplicate as defun in postscript backend (they are different packages, so there won't be collision).